src/query/api/v1/handler/prometheus/remote/write.go

```diff
@@ -590,15 +590,41 @@ func newPromTSIter(timeseries []prompb.TimeSeries, tagOpts models.TagOptions) (*
 
 type promTSIter struct {
 	idx        int
+	err        error
 	attributes []ts.SeriesAttributes
 	tags       []models.Tags
 	datapoints []ts.Datapoints
 	metadatas  []ts.Metadata
+	annotation []byte
 }
 
 func (i *promTSIter) Next() bool {
+	if i.err != nil {
+		return false
+	}
+
 	i.idx++
-	return i.idx < len(i.tags)
+	if i.idx >= len(i.tags) {
+		return false
+	}
+
+	annotationPayload, err := storage.SeriesAttributesToAnnotationPayload(i.attributes[i.idx])
+	if err != nil {
+		i.err = err
+		return false
+	}
+
+	i.annotation, err = annotationPayload.Marshal()
```

Comment:
Should this use `MarshalAppend` rather than `Marshal`? `Marshal` always allocs a slice, but we can pass in a zeroed `i.annotation` into `MarshalAppend` to avoid that

Reply:
I tried using `MarshalTo(data)` initially, but later discovered that this leads to concurrency issues because the output of `promTSIter` is used within multiple goroutines.